File: ga/4omini_result_2024_7/result/1968D_cpp_mini/40_1968D.cpp
use std::io::{self, Read};

fn calculate(p: &[usize], a: &[i64], pos: usize, k: usize) -> i64 {
    let n = p.len();
    let mut mx = 0;
    let mut cur = 0;
    let mut vis = vec![false; n];
    let mut current_pos = pos;
    let mut steps = k;

    while !vis[current_pos] && steps > 0 {
        vis[current_pos] = true;
        mx = mx.max(cur + (steps as i64 * a[current_pos]));
        cur += a[current_pos];
        current_pos = p[current_pos];
        steps -= 1;
    }
    
    mx
}

fn main() {
    let mut input = String::new();
    io::stdin().read_to_string(&mut input).unwrap();
    let mut lines = input.lines();

    let t: usize = lines.next().unwrap().parse().unwrap();
    
    for _ in 0..t {
        let (n, k, ps1, ps2): (usize, usize, usize, usize) = {
            let line = lines.next().unwrap();
            let mut it = line.split_whitespace();
            (
                it.next().unwrap().parse().unwrap(),
                it.next().unwrap().parse().unwrap(),
                it.next().unwrap().parse().unwrap(),
                it.next().unwrap().parse().unwrap(),
            )
        };

        let p: Vec<usize> = lines.next().unwrap().split_whitespace()
            .map(|x| x.parse::<usize>().unwrap() - 1) // converting to 0-based indexing
            .collect();
        
        let a: Vec<i64> = lines.next().unwrap().split_whitespace()
            .map(|x| x.parse::<i64>().unwrap())
            .collect();

        let ans1 = calculate(&p, &a, ps1 - 1, k);
        let ans2 = calculate(&p, &a, ps2 - 1, k);
        
        if ans1 == ans2 {
            println!("Draw");
        } else if ans1 > ans2 {
            println!("Bodya");
        } else {
            println!("Sasha");
        }
    }
}

File: ga/4omini_result_2024_7/result/1968D_mini/40_1968D.rs
use std::io;

const N: usize = 200000;

fn min(a: i32, b: i32) -> i32 {
    if a < b { a } else { b }
}

fn maxl(a: i64, b: i64) -> i64 {
    if a > b { a } else { b }
}

fn main() {
    let stdin = io::stdin();
    let mut input = String::new();
    
    stdin.read_line(&mut input).unwrap();
    let t: usize = input.trim().parse().unwrap();
    
    for _ in 0..t {
        let mut p1 = 0;
        let mut p2 = 0;
        let mut p1_ = 0;
        let mut p2_ = 0;
        let mut i = -1;
        let mut j = -1;
        
        input.clear();
        stdin.read_line(&mut input).unwrap();
        let nums: Vec<i32> = input.trim().split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();
        
        let n = nums[0];
        let k = nums[1];
        let mut a = nums[2] as usize;
        let mut b = nums[3] as usize;
        let m = min(n, k);
        
        let mut arr = vec![0; N + 1];
        let mut aa = vec![0; N + 1];
        
        input.clear();
        stdin.read_line(&mut input).unwrap();
        for (i, num) in input.trim().split_whitespace().enumerate() {
            arr[i + 1] = num.parse().unwrap();
        }
        
        input.clear();
        stdin.read_line(&mut input).unwrap();
        for (i, num) in input.trim().split_whitespace().enumerate() {
            aa[i + 1] = num.parse().unwrap();
        }
        
        for x in 1..=m {
            p1 = maxl(p1, (aa[a] as i64) * (k - x + 1) + p1_);
            p1_ += aa[a];
            a = arr[a] as usize;
            
            p2 = maxl(p2, (aa[b] as i64) * (k - x + 1) + p2_);
            p2_ += aa[b];
            b = arr[b] as usize;
        }
        
        if p1 > p2 {
            println!("Bodya");
        } else if p1 < p2 {
            println!("Sasha");
        } else {
            println!("Draw");
        }
    }
}